File: src/MagnusAddon/note/vocabulary/vocabnote_meta_tag.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots
from language_services.janome_ex.tokenizing.pre_processing_stage.ichidan_godan_potential_or_imperative_hybrid_splitter import IchidanGodanPotentialOrImperativeHybridSplitter
from language_services.janome_ex.tokenizing.pre_processing_stage.word_info_entry import WordInfoEntry
from note.tags import Tags
from note.vocabulary.pos import POS
from typed_linq_collections.collections.q_set import QSet

if TYPE_CHECKING:
    from note.vocabulary.vocabnote import VocabNote

class VocabMetaTag(Slots):
    def __init__(self, name: str, display: str, tooltip: str) -> None:
        self.name: str = name
        self.display: str = display
        self.tooltip: str = tooltip

def get_meta_tags_html(vocab: VocabNote, display_extended_sentence_statistics: bool = True, no_sentense_statistics: bool = False) -> str:
    tags = vocab.tags
    tag_names = vocab.tags.select(lambda it: it.name).to_set()
    meta: list[VocabMetaTag] = []
    tos = QSet(t.lower().strip() for t in vocab.parts_of_speech.raw_string_value().split(","))

    if not no_sentense_statistics:
        def max_nine_number(value: int) -> str: return f"""{value}""" if value < 10 else "+"
        highlighted_in = vocab.sentences.user_highlighted()
        meta.append(VocabMetaTag("highlighted_in_sentences", f"""{max_nine_number(len(highlighted_in))}""", f"""highlighted in {len(highlighted_in)} sentences"""))

        counts = vocab.sentences.counts()
        if counts.total > 0:
            tooltip_text = f"""in {counts.total} sentences. Studying-listening:{counts.studying_listening}, Studying-reading:{counts.studying_reading}"""
            if counts.studying_listening > 0 or counts.studying_reading > 0:
                if display_extended_sentence_statistics:
                    meta.append(VocabMetaTag("in_studying_sentences", f"""{counts.studying_listening}:{counts.studying_reading}/{counts.total}""", tooltip_text))
                else:
                    def create_display_text() -> str:
                        if counts.studying_listening > 9 and counts.studying_reading > 9:
                            return "+"
                        return f"{max_nine_number(counts.studying_listening)}:{max_nine_number(counts.studying_reading)}/{max_nine_number(counts.total)}"

                    meta.append(VocabMetaTag("in_studying_sentences", create_display_text(), tooltip_text))
            else:
                meta.append(VocabMetaTag("in_sentences", f"""{counts.total}""", tooltip_text))
        else:
            meta.append(VocabMetaTag("in_no_sentences", f"""{counts.total}""", f"""in {counts.total} sentences"""))

    # overarching info
    if "_uk" in tag_names: meta.append(VocabMetaTag("uk", "uk", "usually written using kana only"))
    if POS.EXPRESSION in tos: meta.append(VocabMetaTag(POS.EXPRESSION, "x", "expression"))
    if POS.ABBREVIATION in tos: meta.append(VocabMetaTag("abbreviation", "abbr", "abbreviation"))
    if POS.AUXILIARY in tos: meta.append(VocabMetaTag("auxiliary", "aux", "auxiliary"))
    if POS.PREFIX in tos: meta.append(VocabMetaTag("prefix", "頭", "prefix"))
    if POS.SUFFIX in tos: meta.append(VocabMetaTag("suffix", "尾", "suffix"))

    # nouns
    if POS.PROPER_NOUN in tos: meta.append(VocabMetaTag("proper-noun", "p-名", "proper noun"))
    if POS.PRONOUN in tos: meta.append(VocabMetaTag("pronoun", "pr-名", "pronoun"))
    if POS.NOUN in tos: meta.append(VocabMetaTag(POS.NOUN, "名", "noun"))
    if POS.ADVERBIAL_NOUN in tos: meta.append(VocabMetaTag("adverbial-noun", "副-名", "adverbial noun"))
    if POS.INDEPENDENT_NOUN in tos: meta.append(VocabMetaTag("independent-noun", "i-名", "independent noun"))

    # verbs
    if POS.ICHIDAN_VERB in tos: meta.append(_create_verb_meta_tag("ichidan", "1", POS.ICHIDAN_VERB, tos))
    if POS.GODAN_VERB in tos: meta.append(_create_verb_meta_tag("godan", "5", POS.GODAN_VERB, tos))
    if POS.SURU_VERB in tos or "verbal noun" in tos or "する verb" in tos: meta.append(_create_verb_meta_tag("suru-verb", "為", POS.SURU_VERB, tos))
    if POS.KURU_VERB in tos: meta.append(_create_verb_meta_tag("kuru-verb", "k-v", "kuru verb", tos))
    if "auxiliary verb" in tos: meta.append(_create_verb_meta_tag("auxiliary-verb", "aux-v", "auxiliary verb", tos))

    # adverbs
    if POS.TO_ADVERB in tos: meta.append(VocabMetaTag("to-adverb", "と", "adverbial noun taking the と particle to act as adverb"))
    elif POS.ADVERB in tos: meta.append(VocabMetaTag("adverb", "副", "adverb"))
    elif POS.ADVERBIAL in tos: meta.append(VocabMetaTag("adverbial", "副", "adverbial"))

    # adjectives
    if POS.I_ADJECTIVE in tos: meta.append(VocabMetaTag("i-adjective", "い", "true adjective ending on the い copula"))
    if POS.NA_ADJECTIVE in tos: meta.append(VocabMetaTag("na-adjective", "な", "adjectival noun taking the な particle to act as adjective"))
    if POS.NO_ADJECTIVE in tos: meta.append(VocabMetaTag("no-adjective", "の", "adjectival noun taking the の particle to act as adjective"))
    if "auxiliary adjective" in tos: meta.append(VocabMetaTag("auxiliary-adjective", "a-い", "auxiliary adjective"))

    # misc
    if POS.COUNTER in tos: meta.append(VocabMetaTag("counter", "ctr", "counter"))
    if POS.NUMERAL in tos: meta.append(VocabMetaTag("numeral", "num", "numeral"))
    if POS.INTERJECTION in tos: meta.append(VocabMetaTag("interjection", "int", "interjection"))
    if POS.CONJUNCTION in tos: meta.append(VocabMetaTag("conjunction", "conj", "conjunction"))
    if POS.PARTICLE in tos: meta.append(VocabMetaTag("particle", "prt", "particle"))

    # my own inventions
    if POS.MASU_SUFFIX in tos: meta.append(VocabMetaTag("masu-suffix", "連", "follows the 連用形/masu-stem form of a verb"))

    # register
    if Tags.Vocab.Register.polite in tags: meta.append(VocabMetaTag("register-polite", "P", "Polite"))
    if Tags.Vocab.Register.formal in tags: meta.append(VocabMetaTag("register-formal", "F", "Formal"))
    if Tags.Vocab.Register.informal in tags: meta.append(VocabMetaTag("register-informal", "I", "Informal"))
    if Tags.Vocab.Register.archaic in tags: meta.append(VocabMetaTag("register-archaic", "A", "Archaic"))
    if Tags.Vocab.Register.sensitive in tags: meta.append(VocabMetaTag("register-sensitive", "Se", "Sensitive"))
    if Tags.Vocab.Register.vulgar in tags: meta.append(VocabMetaTag("register-vulgar", "V", "Vulgar, usually offensive"))
    if Tags.Vocab.Register.childish in tags: meta.append(VocabMetaTag("register-childish", "C", "Childish, apt to make the speaker sound immature"))
    if Tags.Vocab.Register.slang in tags: meta.append(VocabMetaTag("register-slang", "Sl", "Slang"))
    if Tags.Vocab.Register.humble in tags: meta.append(VocabMetaTag("register-humble", "Hu", "Humble speech"))
    if Tags.Vocab.Register.honorific in tags: meta.append(VocabMetaTag("register-honorific", "Ho", "Honorific form, used to elevate the listener"))
    if Tags.Vocab.Register.rough_masculine in tags: meta.append(VocabMetaTag("register-rough-masculine", "R", "Rough speech, traditionally thought masculine"))
    if Tags.Vocab.Register.soft_feminine in tags: meta.append(VocabMetaTag("register-soft-feminine", "S", "Soft speech, traditionally thought feminine"))
    if Tags.Vocab.Register.derogatory in tags: meta.append(VocabMetaTag("register-derogatory", "D", "Derogatory form, usually offensive"))
    if Tags.Vocab.Register.literary in tags: meta.append(VocabMetaTag("register-literary", "L", "literary, apt to stand out in speech"))

    # other
    if Tags.Vocab.is_ichidan_hiding_godan_potential in tags:
        hidden_godan: WordInfoEntry | None = IchidanGodanPotentialOrImperativeHybridSplitter.try_get_godan_hidden_by_ichidan(vocab)
        meta.append(VocabMetaTag("is-ichidan-hiding-godan-potential", "HG",
                                 f"""Ichidan verb hiding godan potential form of the verb:
{hidden_godan.word if hidden_godan is not None else "unknown"}:
{hidden_godan.answer if hidden_godan is not None else "unknown"}
Mark the ichidan as an incorrect match to see the godan potential in the breakdown. The parser cannot tell which it is on its own."""))

    return """<ol class="vocab_tag_list">""" + "".join([f"""<li class="vocab_tag vocab_tag_{tag.name}" title="{tag.tooltip}">{tag.display}</li>""" for tag in meta]) + "</ol>"

def _create_verb_meta_tag(name: str, display: str, tooltip: str, tos: QSet[str]) -> VocabMetaTag:
    tag = VocabMetaTag(name, display, tooltip)

    if POS.INTRANSITIVE in tos:
        tag.display += "i"
        tag.tooltip = "intransitive " + tag.tooltip
    if POS.TRANSITIVE in tos:
        tag.display += "t"
        tag.tooltip = "transitive " + tag.tooltip

    return tag

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/VocabNoteMetaTag.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using JAStudio.Core.LanguageServices.JanomeEx.Tokenizing.PreProcessingStage;

namespace JAStudio.Core.Note.Vocabulary;

public class VocabMetaTag
{
    public string Name { get; set; }
    public string Display { get; set; }
    public string Tooltip { get; set; }

    public VocabMetaTag(string name, string display, string tooltip)
    {
        Name = name;
        Display = display;
        Tooltip = tooltip;
    }
}

public static class VocabNoteMetaTagFormatter
{
    public static string GetMetaTagsHtml(VocabNote vocab, bool displayExtendedSentenceStatistics = true, bool noSentenceStatistics = false)
    {
        var tags = vocab.Tags;
        var tagNames = tags.Select(t => t.Name).ToHashSet();
        var meta = new List<VocabMetaTag>();
        var tos = vocab.PartsOfSpeech.RawStringValue()
            .Split(',')
            .Select(t => t.ToLower().Trim())
            .ToHashSet();

        if (!noSentenceStatistics)
        {
            string MaxNineNumber(int value) => value < 10 ? value.ToString() : "+";
            
            var highlightedIn = vocab.Sentences.UserHighlighted();
            meta.Add(new VocabMetaTag("highlighted_in_sentences", MaxNineNumber(highlightedIn.Count), $"highlighted in {highlightedIn.Count} sentences"));

            var counts = vocab.Sentences.Counts();
            if (counts.Total > 0)
            {
                var tooltipText = $"in {counts.Total} sentences. Studying-listening:{counts.StudyingListening}, Studying-reading:{counts.StudyingReading}";
                if (counts.StudyingListening > 0 || counts.StudyingReading > 0)
                {
                    if (displayExtendedSentenceStatistics)
                    {
                        meta.Add(new VocabMetaTag("in_studying_sentences", $"{counts.StudyingListening}:{counts.StudyingReading}/{counts.Total}", tooltipText));
                    }
                    else
                    {
                        string CreateDisplayText()
                        {
                            if (counts.StudyingListening > 9 && counts.StudyingReading > 9)
                                return "+";
                            return $"{MaxNineNumber(counts.StudyingListening)}:{MaxNineNumber(counts.StudyingReading)}/{MaxNineNumber(counts.Total)}";
                        }
                        meta.Add(new VocabMetaTag("in_studying_sentences", CreateDisplayText(), tooltipText));
                    }
                }
                else
                {
                    meta.Add(new VocabMetaTag("in_sentences", counts.Total.ToString(), tooltipText));
                }
            }
            else
            {
                meta.Add(new VocabMetaTag("in_no_sentences", counts.Total.ToString(), $"in {counts.Total} sentences"));
            }
        }

        // Overarching info
        if (tagNames.Contains("_uk")) meta.Add(new VocabMetaTag("uk", "uk", "usually written using kana only"));
        if (tos.Contains(POS.Expression)) meta.Add(new VocabMetaTag(POS.Expression, "x", "expression"));
        if (tos.Contains(POS.Abbreviation)) meta.Add(new VocabMetaTag("abbreviation", "abbr", "abbreviation"));
        if (tos.Contains(POS.Auxiliary)) meta.Add(new VocabMetaTag("auxiliary", "aux", "auxiliary"));
        if (tos.Contains(POS.Prefix)) meta.Add(new VocabMetaTag("prefix", "頭", "prefix"));
        if (tos.Contains(POS.Suffix)) meta.Add(new VocabMetaTag("suffix", "尾", "suffix"));

        // Nouns
        if (tos.Contains(POS.ProperNoun)) meta.Add(new VocabMetaTag("proper-noun", "p-名", "proper noun"));
        if (tos.Contains(POS.Pronoun)) meta.Add(new VocabMetaTag("pronoun", "pr-名", "pronoun"));
        if (tos.Contains(POS.Noun)) meta.Add(new VocabMetaTag(POS.Noun, "名", "noun"));
        if (tos.Contains(POS.AdverbialNoun)) meta.Add(new VocabMetaTag("adverbial-noun", "副-名", "adverbial noun"));
        if (tos.Contains(POS.IndependentNoun)) meta.Add(new VocabMetaTag("independent-noun", "i-名", "independent noun"));

        // Verbs
        if (tos.Contains(POS.IchidanVerb)) meta.Add(CreateVerbMetaTag("ichidan", "1", POS.IchidanVerb, tos));
        if (tos.Contains(POS.GodanVerb)) meta.Add(CreateVerbMetaTag("godan", "5", POS.GodanVerb, tos));
        if (tos.Contains(POS.SuruVerb) || tos.Contains("verbal noun") || tos.Contains("する verb")) meta.Add(CreateVerbMetaTag("suru-verb", "為", POS.SuruVerb, tos));
        if (tos.Contains(POS.KuruVerb)) meta.Add(CreateVerbMetaTag("kuru-verb", "k-v", "kuru verb", tos));
        if (tos.Contains("auxiliary verb")) meta.Add(CreateVerbMetaTag("auxiliary-verb", "aux-v", "auxiliary verb", tos));

        // Adverbs
        if (tos.Contains(POS.ToAdverb)) meta.Add(new VocabMetaTag("to-adverb", "と", "adverbial noun taking the と particle to act as adverb"));
        else if (tos.Contains(POS.Adverb)) meta.Add(new VocabMetaTag("adverb", "副", "adverb"));
        else if (tos.Contains(POS.Adverbial)) meta.Add(new VocabMetaTag("adverbial", "副", "adverbial"));

        // Adjectives
        if (tos.Contains(POS.IAdjective)) meta.Add(new VocabMetaTag("i-adjective", "い", "true adjective ending on the い copula"));
        if (tos.Contains(POS.NaAdjective)) meta.Add(new VocabMetaTag("na-adjective", "な", "adjectival noun taking the な particle to act as adjective"));
        if (tos.Contains(POS.NoAdjective)) meta.Add(new VocabMetaTag("no-adjective", "の", "adjectival noun taking the の particle to act as adjective"));
        if (tos.Contains("auxiliary adjective")) meta.Add(new VocabMetaTag("auxiliary-adjective", "a-い", "auxiliary adjective"));

        // Misc
        if (tos.Contains(POS.Counter)) meta.Add(new VocabMetaTag("counter", "ctr", "counter"));
        if (tos.Contains(POS.Numeral)) meta.Add(new VocabMetaTag("numeral", "num", "numeral"));
        if (tos.Contains(POS.Interjection)) meta.Add(new VocabMetaTag("interjection", "int", "interjection"));
        if (tos.Contains(POS.Conjunction)) meta.Add(new VocabMetaTag("conjunction", "conj", "conjunction"));
        if (tos.Contains(POS.Particle)) meta.Add(new VocabMetaTag("particle", "prt", "particle"));

        // My own inventions
        if (tos.Contains(POS.MasuSuffix)) meta.Add(new VocabMetaTag("masu-suffix", "連", "follows the 連用形/masu-stem form of a verb"));

        // Register
        if (tags.Contains(Tags.Vocab.Register.Polite)) meta.Add(new VocabMetaTag("register-polite", "P", "Polite"));
        if (tags.Contains(Tags.Vocab.Register.Formal)) meta.Add(new VocabMetaTag("register-formal", "F", "Formal"));
        if (tags.Contains(Tags.Vocab.Register.Informal)) meta.Add(new VocabMetaTag("register-informal", "I", "Informal"));
        if (tags.Contains(Tags.Vocab.Register.Archaic)) meta.Add(new VocabMetaTag("register-archaic", "A", "Archaic"));
        if (tags.Contains(Tags.Vocab.Register.Sensitive)) meta.Add(new VocabMetaTag("register-sensitive", "Se", "Sensitive"));
        if (tags.Contains(Tags.Vocab.Register.Vulgar)) meta.Add(new VocabMetaTag("register-vulgar", "V", "Vulgar, usually offensive"));
        if (tags.Contains(Tags.Vocab.Register.Childish)) meta.Add(new VocabMetaTag("register-childish", "C", "Childish, apt to make the speaker sound immature"));
        if (tags.Contains(Tags.Vocab.Register.Slang)) meta.Add(new VocabMetaTag("register-slang", "Sl", "Slang"));
        if (tags.Contains(Tags.Vocab.Register.Humble)) meta.Add(new VocabMetaTag("register-humble", "Hu", "Humble speech"));
        if (tags.Contains(Tags.Vocab.Register.Honorific)) meta.Add(new VocabMetaTag("register-honorific", "Ho", "Honorific form, used to elevate the listener"));
        if (tags.Contains(Tags.Vocab.Register.RoughMasculine)) meta.Add(new VocabMetaTag("register-rough-masculine", "R", "Rough speech, traditionally thought masculine"));
        if (tags.Contains(Tags.Vocab.Register.SoftFeminine)) meta.Add(new VocabMetaTag("register-soft-feminine", "S", "Soft speech, traditionally thought feminine"));
        if (tags.Contains(Tags.Vocab.Register.Derogatory)) meta.Add(new VocabMetaTag("register-derogatory", "D", "Derogatory form, usually offensive"));
        if (tags.Contains(Tags.Vocab.Register.Literary)) meta.Add(new VocabMetaTag("register-literary", "L", "literary, apt to stand out in speech"));

        // other
        if (tags.Contains(Tags.Vocab.IsIchidanHidingGodanPotential))
        {
            var hiddenGodan = IchidanGodanPotentialOrImperativeHybridSplitter.TryGetGodanHiddenByIchidan(vocab);
            var word = hiddenGodan?.Word ?? "unknown";
            var answer = hiddenGodan?.Answer ?? "unknown";
            meta.Add(new VocabMetaTag("is-ichidan-hiding-godan-potential", "HG",
                $"Ichidan verb hiding godan potential form of the verb:\n{word}:\n{answer}\nMark the ichidan as an incorrect match to see the godan potential in the breakdown. The parser cannot tell which it is on its own."));
        }

        var sb = new StringBuilder("<ol class=\"vocab_tag_list\">");
        foreach (var tag in meta)
        {
            sb.Append($"<li class=\"vocab_tag vocab_tag_{tag.Name}\" title=\"{tag.Tooltip}\">{tag.Display}</li>");
        }
        sb.Append("</ol>");
        
        return sb.ToString();
    }

    private static VocabMetaTag CreateVerbMetaTag(string name, string display, string tooltip, HashSet<string> tos)
    {
        var tag = new VocabMetaTag(name, display, tooltip);

        if (tos.Contains(POS.Intransitive))
        {
            tag.Display += "i";
            tag.Tooltip = "intransitive " + tag.Tooltip;
        }
        if (tos.Contains(POS.Transitive))
        {
            tag.Display += "t";
            tag.Tooltip = "transitive " + tag.Tooltip;
        }

        return tag;
    }
}
